test/server/echo.go

```diff
@@ -120,3 +143,61 @@ func main() {
 	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
 	<-sigs
 }
+
+func setResponseFromCodes(request *http.Request, response http.ResponseWriter) error {
+	responseCodes := request.FormValue("codes");
+
+	codes, err := validateCodes(responseCodes)
+	if err != nil {
+		return err
+	}
+
+	// Keep a cursor for each sequence of response codes
+	position, ok := state[responseCodes]
```

Comment:
Please add a lock. Behavior of concurrent read/write to a map is undefined behavior (see https://blog.golang.org/go-maps-in-action).